stream/src/public_key/connection.rs

```diff
@@ -24,35 +23,57 @@ const ENCRYPTION_TAG_LENGTH: usize = 16;
 /// An incoming connection with a verified peer handshake.
 pub struct IncomingConnection<C: Scheme, Si: Sink, St: Stream> {
     config: Config<C>,
-    handshake: IncomingHandshake<Si, St, C>,
+    sink: Si,
+    stream: St,
+    deadline: SystemTime,
+    ephemeral_public_key: x25519::PublicKey,
+    peer_public_key: C::PublicKey,
 }
 
 impl<C: Scheme, Si: Sink, St: Stream> IncomingConnection<C, Si, St> {
-    /// Verify the handshake of an incoming connection.
-    pub async fn verify<R: Rng + CryptoRng + Spawner + Clock>(
-        context: &R,
+    pub async fn verify<E: Clock + Spawner>(
+        context: &E,
         config: Config<C>,
         sink: Si,
-        stream: St,
+        mut stream: St,
     ) -> Result<Self, Error> {
-        let handshake = IncomingHandshake::verify(
+        // Set handshake deadline
+        let deadline = context.current() + config.handshake_timeout;
+
+        // Wait for up to handshake timeout for response
+        let msg = select! {
+            _ = context.sleep_until(deadline) => { return Err(Error::HandshakeTimeout) },
+            result = recv_frame(&mut stream, config.max_message_size) => { result? },
+        };
+
+        // Verify handshake message from peer
+        let signed_handshake =
```

Comment:
Does this assert that the handshake is exactly the size we expect (no trailing bytes)? Or is the assumption that we just rely on the network layer to ensure we don't get too much junk with any particular message?

(This was the rationale for adding a type to the codec everywhere so we could both parse different objects explicitly + assert those different objects have no trailing bytes).

Reply:
I think this could be accomplished by adding a `.done() -> Result<(), CodecError>` that checks remaining is zero and if not returns an error with bytes left?

Reply:
Yes! `decode` [takes care of the trailing bytes](https://github.com/commonwarexyz/monorepo/blob/5b09315df818cd0c084cc2b8086b708940631eb1/codec/src/codec.rs#L35-L37)!

Reply:
I think it might be worth commenting this under the // Verify... because auditor's might not appreciate/realize this.

SO nice